src/internal/operations/export.go

```diff
@@ -0,0 +1,440 @@
+}
+
+// aggregates stats from the export.Run().
+// primarily used so that the defer can take in a
+// pointer wrapping the values, while those values
+// get populated asynchronously.
+type exportStats struct {
+	cs            []data.RestoreCollection
+	ctrl          *data.CollectionStats
+	bytesRead     *stats.ByteCounter
+	resourceCount int
+
+	// a transient value only used to pair up start-end events.
+	exportID string
+}
+
+// Run begins a synchronous export operation.
+func (op *ExportOperation) Run(ctx context.Context) (
+	expColl []export.Collection,
+	err error,
+) {
+	defer func() {
+		if crErr := crash.Recovery(ctx, recover(), "export"); crErr != nil {
+			err = crErr
+		}
+	}()
+
+	var (
+		opStats = exportStats{
+			bytesRead: &stats.ByteCounter{},
+			exportID:  uuid.NewString(),
+		}
+		start  = time.Now()
+		sstore = streamstore.NewStreamer(op.kopia, op.acct.ID(), op.Selectors.PathService())
+	)
+
+	// -----
+	// Setup
+	// -----
+
+	ctx, end := diagnostics.Span(ctx, "operations:export:run")
+	defer func() {
+		end()
+		// wait for the progress display to clean up
+		observe.Complete()
+	}()
+
+	ctx, flushMetrics := events.NewMetrics(ctx, logger.Writer{Ctx: ctx})
+	defer flushMetrics()
+
+	ctx = clues.Add(
+		ctx,
+		"tenant_id", clues.Hide(op.acct.ID()),
+		"backup_id", op.BackupID,
+		"service", op.Selectors.Service)
+
+	defer func() {
+		op.bus.Event(
+			ctx,
+			events.ExportEnd,
+			map[string]any{
+				events.BackupID:      op.BackupID,
+				events.DataRetrieved: op.Results.BytesRead,
+				events.Duration:      op.Results.CompletedAt.Sub(op.Results.StartedAt),
+				events.EndTime:       dttm.Format(op.Results.CompletedAt),
+				events.ItemsRead:     op.Results.ItemsRead,
+				events.ItemsWritten:  op.Results.ItemsWritten,
+				events.Resources:     op.Results.ResourceOwners,
+				events.ExportID:      opStats.exportID,
+				events.Service:       op.Selectors.Service.String(),
+				events.StartTime:     dttm.Format(op.Results.StartedAt),
+				events.Status:        op.Status.String(),
+			})
+	}()
+
+	// -----
+	// Execution
+	// -----
+
+	expCollections, err := op.do(ctx, &opStats, sstore, start)
+	if err != nil {
+		// No return here!  We continue down to persistResults, even in case of failure.
+		logger.CtxErr(ctx, err).Error("running export")
+
+		if errors.Is(err, kopia.ErrNoRestorePath) {
+			op.Errors.Fail(clues.New("empty backup or unknown path provided"))
+		}
+
+		op.Errors.Fail(clues.Wrap(err, "running export"))
+	}
+
+	finalizeErrorHandling(ctx, op.Options, op.Errors, "running export")
+	LogFaultErrors(ctx, op.Errors.Errors(), "running export")
+
+	// -----
+	// Persistence
+	// -----
+
+	err = op.persistResults(ctx, start, &opStats)
+	if err != nil {
+		op.Errors.Fail(clues.Wrap(err, "persisting export results"))
+		return nil, op.Errors.Failure()
+	}
+
+	logger.Ctx(ctx).Infow("completed export", "results", op.Results)
+
+	return expCollections, nil
+}
+
+type zipExportCol struct {
```

Comment:
I think the code will turn out much clearner in the long term if we move anything related to archive processing into its own file, as a way to separate concerns.  So, `/export/zip.go`, in this case. 

Ideally what I'd like to see is for archival to have its own package in internal, completely separate from the export package.  Something like `internal/archive/zip.go` with `package archive`, or some such.  That would more closely follow our current implementations with kopia and m365, as well as what we might imagine from other producers and consumers.  

Separately, this name feels like it's dodging a conflict with the func `zipExportCollection`, and I think it's coming out worse for that (`col` as a suffix, specifically.)  Another benefit of moving this to an internal package and replicating current patterns: instead of having `zip` specific functions and collections: make a plain `Collection` struct in the archive package; let that collection define its methodology / categorization (zip, in this case); centralize processing around standard package behavior while separating per-category specifics into different handlers.

Reply:
I've created a separate package under `internal/archive`. It still work on `export.Collection` instead of `data.Collection` as this still runs as the very last step. That said, if we want to change it in future if we decide to as the implementation is not exposed to the consumer.